### PR TITLE
#166272107 user should be able to create articles

### PR DESCRIPTION
**Pull request by Sojisoyoye**
#### What does this PR do?

This PR is for user should be able to create articles

#### Description of Task to be completed?

- create an association between user and article  model
- create article controller
- define post create article route
- write the test for creating article 

#### How should this be manually tested?

- cd into ft-create-article-166272107
- run `npm install`
- run `npm run start-dev`
- sign up as a user on url http://localhost:3000/api/v1/auth/signup on postman
- a cookie will be generated for that user, then
- go to url http://localhost:3000/api/v1/article on postman
- fill in title, body
- click on send 

#### Any background context you want to provide?

This is to enable a user to create articles

#### What are the relevant pivotal tracker stories?

 [166272107](https://www.pivotaltracker.com/n/projects/2351184/stories/166272107)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -36,7 +36,8 @@
     "request": "^2.87.0",
     "sequelize": "^5.8.6",
     "sequelize-cli": "^5.4.0",
-    "underscore": "^1.9.1"
+    "underscore": "^1.9.1",
```
Comment (EmmanuelSage) [inline package.json]: Hello Soji, would you please help remove `underscore`

File: server/controllers/articleController.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import uniqid from 'uniqid';
+import articles from '../models/articles';
+
+
+/**
+ * @class ArticleController
+ * @exports ArticleController
```
Comment (EmmanuelSage) [inline server/controllers/articleController.js]: Hello Soji, would you mind renaming this to just `article.js`, as it is already in the controller folder?

File: server/controllers/article.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+const { Article } = models;
+
+const createArticle = async (req, res) => {
+  const { title, body } = req.body;
+
+  const { userId } = req.query;
+
+  try {
+    const newArticle = await Article.create({ title, body, userId });
```
Comment (mosinmiloluwa01) [inline server/controllers/article.js]: I think it'll be better if the user id was not passed manually

File: server/fixtures/user.js [hunk 4]
```diff
@@ -27,3 +27,20 @@ export const getUserData = args => ({
   ...goodUserData,
   ...args
 });
+export const firstUser = {
```
Comment (mosinmiloluwa01) [inline server/fixtures/user.js]: please make use of the getUserData function for your objects

File: server/routes/index.js [hunk 5]
```diff
@@ -3,6 +3,6 @@ import indexRouter from './v1';
 
 const router = express.Router();
 
-router.use('/api/v1', indexRouter);
+router.use('/api/v1/', indexRouter);
```
Comment (mosinmiloluwa01) [inline server/routes/index.js]: please remove the / after v1

File: server/routes/v1/article.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import express from 'express';
+import Article from '<controllers>/article';
```
Comment (mosinmiloluwa01) [inline server/routes/v1/article.js]: you might want to change this to article instead of Article

File: server/tests/controller/article.spec.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      article.userId = user.id;
+    } catch (error) {
+      console.log(error);
+    }
+  });
+
+  it('should create a new article controller', (done) => {
+    chai
+      .request(app)
+      .post(`/api/v1/articles?userId=${article.userId}`)
```
Comment (mosinmiloluwa01) [inline server/tests/controller/article.spec.js]: you might also want to modify your tests as well. I dont think there is a need to manually but user id

File: server/tests/models/article.spec.js [hunk 8]
```diff
@@ -1,23 +1,28 @@
 /* eslint-disable no-console */
 import chai from 'chai';
+import chaiHttp from 'chai-http';
 import chaiAsPromise from 'chai-as-promised';
-import models from '../../models';
-import { newArticle } from '../../fixtures/article';
+import models from '<server>/models';
+import { newArticle } from '<fixtures>/article';
+import { firstUser } from '<fixtures>/user';
```
Comment (mosinmiloluwa01) [inline server/tests/models/article.spec.js]: Please use getuserData function instead

File: server/tests/models/article.spec.js [hunk 9]
```diff
@@ -1,23 +1,28 @@
...
 chai.use(chaiAsPromise);
 const { expect } = chai;
-const { Article, sequelize } = models;
+chai.use(chaiHttp);
+const { Article, User, sequelize } = models;
 
 before(async () => {
   try {
     await sequelize.sync({ force: true });
+    const user = await User.create(firstUser);
```
Comment (mosinmiloluwa01) [inline server/tests/models/article.spec.js]: you can call the getUserData function and then pass in the new properties you want to test for

File: server/tests/controller/article.spec.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+before(async () => {
+  try {
+    await sequelize.sync({ force: true });
+  } catch (error) {
+    console.log(error);
+  }
+});
+
+describe('/POST create an article', async () => {
```
Comment (mosinmiloluwa01) [inline server/tests/controller/article.spec.js]: please remove the /post from here

File: server/tests/controller/article.spec.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  before(async () => {
+    try {
+      const user = await User.create(secondUser);
```
Comment (mosinmiloluwa01) [inline server/tests/controller/article.spec.js]: same applies here... please use getUserData function

File: server/tests/controller/article.spec.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+      .request(app)
+      .post(`/api/v1/articles?userId=${article.userId}`)
+      .send(article)
+      .end((err, res) => {
+        expect(res).to.have.status(201);
+        expect(res.body.status).to.be.equal(201);
+        done();
+      });
+  });
+});
```
Comment (mosinmiloluwa01) [inline server/tests/controller/article.spec.js:58]: I also think it will be nice to test for scenarios where the user enters an empty object or invalid fields

File: package.json [hunk 13]
```diff
@@ -69,8 +71,9 @@
   },
   "nyc": {
     "exclude": [
+      "server/tests",
       "server/models/index.js",
-      "server/tests"
+      "server/middlewares/*.js"
```
Comment (mosinmiloluwa01) [inline package.json]: I don't think it is a good idea to exclude middlewares from your coverage.

File: server/controllers/article.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
+import models from '<server>/models';
+
+const { Article } = models;
+
+const createArticle = async (req, res) => {
+  const { title, body } = req.body;
+
+  const { userId } = req.query;
```
Comment (mosinmiloluwa01) [inline server/controllers/article.js]: you might want to consider getting the userId from a middleware

File: server/controllers/article.js [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+
+  try {
+    const newArticle = await Article.create({ title, body, userId });
+
+    res.status(201).json({
+      status: 201,
```
Comment (mosinmiloluwa01) [inline server/controllers/article.js]: Benny said we should use success instead of 201

File: server/fixtures/article.js [hunk 16]
```diff
@@ -5,11 +5,15 @@ const slugTitle = createUniqueSlug('hello world');
...
-  userId: 'a5b37ee6-1553-4681-a343-d8296344e129',
+  userId: 'f1e2e1c8-6a90-4bcd-b349-49f989d311d6',
   isBlacklisted: false,
   slug: slugTitle,
   categoryId: '325163a1-6e8c-4cd8-b5d3-7019c49983f5',
   isPublished: true
 };
 
-export default { newArticle };
+export const article = {
```
Comment (mosinmiloluwa01) [inline server/fixtures/article.js:34]: i think you can use newArticle instead of creating another article

File: server/tests/controller/article.spec.js [hunk 17]
```diff
@@ -0,0 +1,43 @@
+/* eslint-disable no-console */
+import chai from 'chai';
+import chaiHttp from 'chai-http';
+import app from '<server>/app';
+import models from '<server>/models';
+import { article } from '<fixtures>/article';
```
Comment (mosinmiloluwa01) [inline server/tests/controller/article.spec.js:8]: i think it'll be better to use the newArticle object instead of creating a new one

File: server/tests/controller/article.spec.js [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+import { secondUser } from '<fixtures>/user';
```
Comment (mosinmiloluwa01) [inline server/tests/controller/article.spec.js]: Please use getUserData function instead

**Review by mosinmiloluwa01 — CHANGES_REQUESTED** [review]

File: package.json [hunk 19]
```diff
@@ -4,7 +4,7 @@
   "description": "A Social platform for the creative at heart",
   "main": "app.js",
   "scripts": {
-    "test": "NODE_ENV=test nyc --reporter=html --reporter=text mocha --require @babel/register ./server/tests/index.spec.js --exit",
+    "test": "NODE_ENV=test nyc --reporter=html --reporter=text mocha ./server/tests/index.spec.js --require @babel/register --timeout 30000 --exit",
```
Comment (Eazybee) [inline package.json]: why are you using `--timeout 3000`? I feel like you don't need it.

File: server/controllers/article.js (hunk 14 above)
Comment (Eazybee) [reply server/controllers/article.js]: I think what @mosinmiloluwa01  is trying to say, you should protect the route and `userId` shouldn't be passed in through `req.query`

File: server/controllers/article.js [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+      message: 'Article created successfully',
+      data: newArticle
+    });
+  } catch (error) {
+    return error;
```
Comment (Eazybee) [inline server/controllers/article.js]: I don't think you want to `return error` because if there is an error, a user won't get a response and will be left wondering what's happening 

You can do something like 
```
res.status(500).json({
  status: 'error',
  error,
});
```

File: server/models/article.js [hunk 21]
```diff
@@ -48,6 +48,7 @@ const article = (sequelize, DataTypes) => {
       categoryId: {
         type: DataTypes.UUID,
         allowNull: false,
+        defaultValue: DataTypes.UUIDV4,
```
Comment (Eazybee) [inline server/models/article.js]: same here

File: server/migrations/article.js [hunk 22]
```diff
@@ -20,15 +20,21 @@ const Articles = {
...
     },
     isBlacklisted: {
       type: Sequelize.BOOLEAN,
       allowNull: false
     },
     categoryId: {
       type: Sequelize.UUID,
-      allowNull: false
+      allowNull: false,
+      defaultValue: Sequelize.UUIDV4
```
Comment (Eazybee) [inline server/migrations/article.js]: Please remove `defaultValue: Sequelize.UUIDV4`

**Review by Eazybee — CHANGES_REQUESTED** [review]
Good job done @Sojisoyoye .
Please take a minute to reflect on the feedback

**Review by akinyeleolat — APPROVED** [review]
Good job @Sojisoyoye, would you mind sync with @mosinmiloluwa01 because the ```verifyUser ``` function on your branch is similar to ```authentication``` on her branch. Apart from this, nothing more.

File: server/middlewares/verifyUser.js [hunk 23]
```diff
@@ -0,0 +1,11 @@
+import jwt from 'jsonwebtoken';
+
+const verifyUser = (req, res, next) => {
+  const { token } = req.cookies.accessToken;
+
+  const decoded = jwt.verify(token, process.env.SECRET);
+  req.user = { userId: decoded.id };
```
Comment (frankly034) [inline server/middlewares/verifyUser.js]: The choice of req.user might be misleading as the token contains the user's id; userId might be more intuitive.

File: server/models/article.js [hunk 24]
```diff
@@ -68,8 +68,14 @@ const article = (sequelize, DataTypes) => {
     newArticle.setDataValue('slug', createUniqueSlug(newArticle.title));
     newArticle.setDataValue('description', createEllipsis(newArticle.body));
   });
-  Article.associate = () => {
+  Article.associate = (models) => {
     // associations can be defined here
```
Comment (mosinmiloluwa01) [inline server/models/article.js:72]: please remove comment

File: server/models/user.js [hunk 25]
```diff
@@ -89,8 +89,13 @@ const user = (sequelize, DataTypes) => {
     newUser.password = hashPassword(newUser.password);
   });
 
-  User.associate = () => {
-    // associations can be defined here
+  User.associate = (models) => {
+    // associations between user and article
```
Comment (mosinmiloluwa01) [inline server/models/user.js:93]: please remove comment

File: server/tests/controller/article.spec.js [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+      });
+  });
+
+  it('should return server error for create article', async () => {
+    const req = { body: {}, user: { userId: 'f1e2e1c8-6a90-4bcd-b349-49f989d311d6' } };
+    const res = {
+      status() {},
+      json() {}
+    };
+
```
Comment (mosinmiloluwa01) [inline server/tests/controller/article.spec.js:52]: I think you should also test the creation of articles when a user is verified and not verified

**Review by mosinmiloluwa01 — CHANGES_REQUESTED** [review]

File: server/middlewares/verifyUser.js [hunk 27]
```diff
@@ -0,0 +1,11 @@
+import jwt from 'jsonwebtoken';
+
+const verifyUser = (req, res, next) => {
```
Comment (mosinmiloluwa01) [inline server/middlewares/verifyUser.js:3]: I think this function should be something like "checkToken" or "decodeToken".
I will also like to suggest that the filename be changed too because the file is actually decoding the token and not verifying the user

**Review by mosinmiloluwa01 — CHANGES_REQUESTED** [review]

**Review by frankly034 — APPROVED** [review]
Looks good to me.

File: server/tests/controller/article.spec.js [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+before(async () => {
+  await sequelize.sync({ force: true });
+});
+
+afterEach(() => sinon.restore());
+
+const agent = chai.request.agent(app);
+
+describe('create an article', async () => {
```
Comment (benfluleck) [inline server/tests/controller/article.spec.js]: `describe('Create an Article')`

File: server/tests/models/article.spec.js [hunk 29]
```diff
@@ -1,24 +1,31 @@
...
 const { expect } = chai;
-const { Article, sequelize } = models;
+chai.use(chaiHttp);
+const { Article, User, sequelize } = models;
 
 before(async () => {
   try {
     await sequelize.sync({ force: true });
+    const user = await User.create(getUserData({
+      email: 'johnndoe@email.com', username: 'johnndoee', firstName: 'john', lastName: 'doe', bio: 'The baddest dude', password: '123456'
```
Comment (benfluleck) [inline server/tests/models/article.spec.js]: What is going on here I thought I told you not to pass in all these values

**Review by benfluleck — CHANGES_REQUESTED** [review]
Feedback left

File: server/routes/v1/article.js [hunk 30]
```diff
@@ -1,8 +1,10 @@
 import express from 'express';
+import articles from '<controllers>/article';
+import verifyUser from '<middlewares>/verifyUser';
 import { validateCreateArticle } from '<validations>/article';
 
 const article = express.Router();
 
-article.post('/', validateCreateArticle);
+article.post('/', validateCreateArticle, verifyUser, articles);
```
Comment (mosinmiloluwa01) [inline server/routes/v1/article.js:8]: I think this should not be verifyUser because the middleware is not verifying the user its actually decoding the token that is to be used in your route

**Review by mosinmiloluwa01 — CHANGES_REQUESTED** [review]

File: server/controllers/article.js [hunk 31]
```diff
@@ -0,0 +1,31 @@
+import models from '<serverModels>';
+import { article } from '<fixtures>/article';
+
+const { Article } = models;
+
+const createArticle = async (req, res) => {
+  const { title, body } = req.body;
+
+  const categoryId = req.body.categoryId || article.categoryId;
```
Comment (Eazybee) [inline server/controllers/article.js:9]: Please add a comment to remove `article.categoryId` in future

File: server/controllers/signup.js [hunk 32]
```diff
@@ -18,17 +18,25 @@ const signUp = async (req, res) => {
...
+
+    return res.status(201).send({
+      status: 'success',
+      message: 'User created successfully',
+      data: userInfo
+    });
+  } catch (error) {
+    res.status(500).json({
+      status: 'failed',
+      message: 'Server error'
```
Comment (Eazybee) [inline server/controllers/signup.js:37]: Please return error back to the user, not  `message: 'Server error'`

File: server/fixtures/article.js (hunk 16 above)
Comment (Eazybee) [reply server/fixtures/article.js:34]: I agree with @mosinmiloluwa01, do you mind implementing the feedback or communicating your reasons before Resolving conversation

**Review by benfluleck — APPROVED** [review]

File: server/middlewares/verifyUser.js (hunk 27 above)
Comment (Eazybee) [reply server/middlewares/verifyUser.js:3]: same here, I agree with @mosinmiloluwa01

File: server/models/article.js (hunk 24 above)
Comment (Eazybee) [reply server/models/article.js:72]: same here. Agree with @mosinmiloluwa01

File: server/models/user.js (hunk 25 above)
Comment (Eazybee) [reply server/models/user.js:93]: same here, the code is self-documented.

**Review by Eazybee — CHANGES_REQUESTED** [review]
